lib/rucio/core/replica.py

```diff
@@ -852,6 +862,16 @@ def _list_replicas(dataset_clause, file_clause, state_clause, show_pfns, schemes
                             if sign and isinstance(sign, list) and sign[0]:
                                 pfn = get_signed_url(service='gcs', operation='read', url=pfn, lifetime=signature_lifetime)
 
+                        # is the file part of an archive and can be extract directly?
+                        if protocol.attributes['scheme'] == 'root' and '%s:%s' % (scope, name) in constituents:
+                            # parent archive can have a different access path! need to run full PFN resolve for it
+                            archive_pfn = list(list_replicas(dids=[constituents['%s:%s' % (scope, name)]],
+                                                             schemes=['root'],
+                                                             rse_expression=rse,
+                                                             client_location=client_location,
+                                                             domain=domain))[0]['pfns'].keys()[0]
+                            pfn = '%s?xrdcl.unzip=%s' % (archive_pfn, name)
```

Comment:
only for scheme == xroot. what about the other schemes ?

Reply:
Only root supports direct extraction, so it only makes sense here.

One option could be to just return the archive.zip for all other protocols, but then the client has to realise that it didn't get the file via root and has to to manually unzip.

Reply:
and What about, the `rucio-download` use case, e.g.,`rucio-download  MyScope.file.zip` ?

Reply:
`list_replicas` will construct the right PFN, for `rucio download` it will be transparent

Reply:
To wrap up: one option is that `list_replicas` gives back nested urls (`http://example.org/package.zip#path=path/to/the/file` or `http://example.org/package.zip#path/to/the/file`) for non root schemes and that `rucio donwload` does the extraction client side.

Reply:
# might be a problem if it is part of the filename?
In any way, it does not need to be nested but can just be part of the dictionary. There is already multiple keys in there, we can add another one.

Reply:
Certain clients like xrdcp might support nested urls directly.  An option of `list_replicas`  for the format of the zip file url can be relevant as well.

Reply:
It might be worthwhile to use URL parsing libraries to parse out the URL, then add the `xrdcl.unzip` option, then serialize it back to a string.

Direct manipulation of URLs as strings always seems to end up in tears (or, at least, multiple `?` and/or inadvertently repeated options).

Reply:
I'm rewriting this part already ;-)